ADApp/pluginSrc/NDPluginCodec.cpp

```diff
@@ -635,6 +634,96 @@ NDArray *decompressBSLZ4(NDArray *input, NDCodecStatus_t *status, char *errorMes
 
 #endif // ifdef HAVE_BITSHUFFLE
 
+#ifdef HAVE_VC
+#include <video_compression.h>
+//extern int H264_compress_default(const char*, char*, int, int, int);
+NDArray *compressH264(NDArray *input, NDCodecStatus_t *status, char *errorMessage)
+{
+    //printf("inside compressH264\n");
+    if (!input->codec.empty()) {
+        sprintf(errorMessage, "Array is already compressed");
+        *status = NDCODEC_WARNING;
+        return NULL;
+    }
+
+    switch (input->dataType) {
+        case NDInt8:
+        case NDUInt8:
+        //case NDInt16:
+        //case NDUInt16:
+            break;
+        default:
+            sprintf(errorMessage, "H264 only supports 8-bit data");
+            *status = NDCODEC_ERROR;
+            //goto failure;
+            return NULL;
+    }
+
+
+
+    NDArrayInfo_t info;
+    input->getInfo(&info);
+    int outputSize = LZ4_compressBound((int)info.totalBytes);
+    NDArray *output = allocArray(input, -1, outputSize);
+
+    if (!output) {
+        sprintf(errorMessage, "Failed to allocate H264 output array");
+        *status = NDCODEC_ERROR;
+        return NULL;
+    }
+
+    //int compSize = LZ4_compress_default((const char*)input->pData, (char*)output->pData, (int)info.totalBytes, outputSize);
+    int x_size = input->dims[0].size;
+    int y_size = input->dims[1].size;
+
+    //printf("x_size %d y_size %d\n", x_size, y_size);
+    //printf("before H264 compress default\n"); 
+    //int compSize = H264_compress_default((const char*)input->pData, (char*)output->pData, (int)x_size, (int)y_size, outputSize);
+    int compSize = H264_compress((const char*)input->pData, (char*)output->pData, (int)x_size, (int)y_size);
+
+    if (compSize <= 0) {
+        output->release();
+        sprintf(errorMessage, "Internal H264 error");
+        *status = NDCODEC_ERROR;
+        return NULL;
+    }
+
+    output->codec.name = codecName[NDCODEC_H264];
+    output->compressedSize = compSize;
+    //printf("before return output\n");
+
+    return output;
+}
+
+#else
+
+NDArray *compressH264(NDArray *input, NDCodecStatus_t *status, char *errorMessage)
+{
+    sprintf(errorMessage, "No H264 support");
+    *status = NDCODEC_ERROR;
+    return NULL;
+}
+
+NDArray *decompressH264(NDArray *input, NDCodecStatus_t *status, char *errorMessage)
+{
+    sprintf(errorMessage, "No H264 support");
+    *status = NDCODEC_ERROR;
+    return NULL;
+}
+
+
+void set_gop_size(int value){
+    printf("unable to set gop size to %d\n", value);
+}
+
+void set_q_min_max(int value){
+    printf("unable to set q min max to %d\n", value);
+}
```

Comment:
These functions are implemented only for the case where `HAVE_VC` isn't defined.

Reply:
I see now that `set_gop_size` and `set_q_min_max` are defined in the ADSupport functions. They are operating on global status; is it safe to operate on them during compression, for example? And this means we can't have more than one compression stream on the same IOC.

Reply:
Yes I think a global variable is not ideal because I believe it is possible to have two separate cameras in a single IOC, and hence multiple instances of the plugin. I did not think of this. In practice, there is usually only camera per IOC. I don't think it is likely that the global variable is the reason for the unlock issues since I did my testing with only one instance of the plugin.

Reply:
> I don't think it is likely that the global variable is the reason for the unlock issues since I did my testing with only one instance of the plugin.

Even with a single instance of the plugin, but without the additional locks you added, if you write into the QMinMax or GOPSize PVs while compression is happening, you will reset the codec state while the struct is being used. That's not safe. These functions should be implemented in NDPluginCodec, which can track the current encoding state, and block writes to these values if encoding is ongoing.